src/Microsoft.ML.Data/Prediction/Calibrator.cs

```diff
@@ -561,12 +563,50 @@ private ParameterMixingCalibratedModelParameters(IHostEnvironment env, ModelLoad
             _featureWeights = SubModel as IPredictorWithFeatureWeights<float>;
         }
 
-        private static CalibratedModelParametersBase Create(IHostEnvironment env, ModelLoadContext ctx)
+        private ParameterMixingCalibratedModelParameters(IHostEnvironment env, ModelLoadContext ctx, TSubModel predictor, TCalibrator calibrator)
+            : base(env, ParameterMixingCalibratedModelParameters.RegistrationName, predictor, calibrator)
+        {
+            Host.Check(SubModel is IParameterMixer<float>, "Predictor does not implement " + nameof(IParameterMixer));
+            Host.Check(SubModel is IPredictorWithFeatureWeights<float>, "Predictor does not implement " + nameof(IPredictorWithFeatureWeights<float>));
+            _featureWeights = SubModel as IPredictorWithFeatureWeights<float>;
+        }
+
+        // ORIGINAL CREATE METHOD
+        //private static CalibratedModelParametersBase Create(IHostEnvironment env, ModelLoadContext ctx)
+        //{
+        //    Contracts.CheckValue(env, nameof(env));
+        //    env.CheckValue(ctx, nameof(ctx));
+        //    ctx.CheckAtModel(GetVersionInfo());
+        //    return new ParameterMixingCalibratedModelParameters<TSubModel, TCalibrator>(env, ctx);
+        //}
+
+        internal static object Create(IHostEnvironment env, ModelLoadContext ctx)
```

Comment:
Do we really need this method and `ParameterMixingCalibratedModelParameters.Create` below? Can't this code just move into `ParameterMixingCalibratedModelParameters.Create`?

Reply:
Well, problem was that this create method calls `GetPredictor()` and `GetCalibrator()` which are static methods of generic class `CMPB<,>`, and it also calls `GetVersionInfo()` which is static method of generic class `PMCMP<,>`.

But yeah, I actually can get rid of this create method, by calling those other methods through `ParameterMixingCalibratedModelParameters<IPredictorProducing<float>, ICalibrator>` inside the `ParameterMixingCalibratedModelParameters.Create` method. I just think it would look uglier, but it's doable.

So I will do this asap. Thanks for the feedback!

Reply:
So I tried to do this as I said, and problem was that GetPredictor(), GetCalibrator() and GetVersionInfo() are all private methods.

There are many ways to refactor this, though... specially since I still have to make changes to the other classes in Calibrator.cs, so I will address this particular comment in my next iteration when I change those other classes.

---
In reply to: [334058843](https://github.com/dotnet/machinelearning/pull/4306#discussion_r334058843) [](ancestors = 334058843)

Reply:
The reason I proposed the above is because of the following code in `ParameterMixingCalibratedModelParameters.Create`:

```C#
public static CalibratedModelParametersBase Create(IHostEnvironment env, ModelLoadContext ctx)
        {
            return (CalibratedModelParametersBase) ParameterMixingCalibratedModelParameters<IPredictorProducing<float>, ICalibrator>.Create(env, ctx);
        }
```

Here we are still using `IPredictorProducing<float>` and `ICalibrator` for the generic types, but I was thinking we were ignoring those types.  I was wrong because we are using those types in `GetPredictor()` and `GetCalibrator()`, as you mention above. So my question was answered. The two separate methods are still necessary. Thanks for checking.